### PR TITLE
Implement missing `#-`, `@?` and `@@` Postgres' JSONB operators.

### LINKED ISSUE #7147: Add missing json/jsonb operators in PostgreSQL
Since the json/jsonb support in SQLAlchemy was added PostgreSQL has added some new operators, like `@@` and `@?` in v12 https://www.postgresql.org/docs/12/datatype-json.html

It would be nice to add the missing operators to the relevant types.

One thing to finalize are the names of these operators, that are not immediately clear from the postgresql docs

First raised in https://github.com/sqlalchemy/sqlalchemy/discussions/7146

### PR DESCRIPTION
**Pull request by gmcrocetti**
### Description

Fixes #7147.
Opening as draft to collect feedback as this is my first time here (:

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by CaselIT — COMMENTED** [review]
Look fine to me.
I don't know if the names are the one usually used by postgresql, since the docs of postgresql do not seem to name them. They look fine to me.

Any reason for not implementing `#-`?

File: lib/sqlalchemy/dialects/postgresql/json.py [hunk 1]
```diff
@@ -340,4 +363,30 @@ def contained_by(self, other):
                 CONTAINED_BY, other, result_type=sqltypes.Boolean
             )
 
+        def deletes_field_or_array_element(self, array):
+            """ JSONB expression. Deletes field or array element specified in
+            the argument array.
+            """
+            return self.operate(
```
Comment (gmcrocetti) [inline lib/sqlalchemy/dialects/postgresql/json.py:370]: @CaselIT . That's the line I need your opinion.
Should we be defensive here and always convert to `sqlalchemy.dialects.postgresql.array` or leave it to the user ? I'm always inclined to the first option but that's a matter of taste/culture of the codebase, which I'm not familiar with.

**Comment by gmcrocetti** [comment]
> Look fine to me. I don't know if the names are the one usually used by postgresql, since the docs of postgresql do not seem to name them. They look fine to me.
> 
> Any reason for not implementing `#-`?

Hey o/. I'm glad it's ok.
Yeap. Briefly went to postgres' codebase but couldn't find anything meaningful.

Regarding the `#-` it's implemented now. I didn't implement it initially due to my lack of familiarity with sqlalchemy's code, tho marked you in a line to collect feedback on how sqlalchemy usually approaches it

File: lib/sqlalchemy/dialects/postgresql/json.py (hunk 1 above)
Comment (CaselIT) [reply lib/sqlalchemy/dialects/postgresql/json.py:371]: I'm not sure about thins. I think it should be possible for an user to pass a simply a list like [1, 'b'] and this should work correctly.

We probably need a round trip tests for this. 

Thoughs @zzzeek ?

**Comment by CaselIT** [comment]
Thanks for the update!

Regarding the names for @@ and @? we seem to have had the names spelled in the issue taken from the postgres codebase.
Regarding `#-` I've looked a bit and the only reference I can find is this one https://github.com/postgres/postgres/blob/33ab0a2a527e3af5beee3a98fc07201e555d6e45/src/include/catalog/pg_operator.dat#L3254-L3256 that seems to suggest that the name of the function is `jsonb_delete_path` (and there seem to be a undocumented function of the same name https://github.com/postgres/postgres/blob/33ab0a2a527e3af5beee3a98fc07201e555d6e45/src/test/regress/sql/jsonb.sql#L1145-L1147

So maybe we could use `delete_path` as the operation name in sqlalchemy? Thoughs @zzzeek ?

For now I'll make it run the pipeline.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 8fc88f466f905c5533601965cc8385b1a9850d1a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 8fc88f466f905c5533601965cc8385b1a9850d1a: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

**Comment by gmcrocetti** [comment]
Nice finding @CaselIT . Just refactored it :)

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 8647aaf2d9f48c55c152673828deb8ed54966a11 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 8647aaf2d9f48c55c152673828deb8ed54966a11 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

The code looks ok, but I think we should add roundtrip tests to #- since the right side uses arrays that can cause issues with different drivers

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

**Comment by CaselIT** [comment]
@gmcrocetti I'll implement the round trip test and add the changelog message. Thanks for the effort so far!

**Comment by gmcrocetti** [comment]
> @gmcrocetti I'll implement the round trip test and add the changelog message. Thanks for the effort so far!

Ok ! I'll keep my eyes close in order to learn :) o/. TY

File: lib/sqlalchemy/dialects/postgresql/json.py [hunk 2]
```diff
@@ -340,4 +364,28 @@ def contained_by(self, other):
...
+        def delete_path(self, array):
+            """JSONB expression. Deletes field or array element specified in
+            the argument array.
+            """
+            return self.operate(DELETE_PATH, array, result_type=JSONB)
+
+        def path_exists(self, other):
+            """Boolean expression. Test for presence of item given by the
+            argument JSONPath expression.
+            """
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/postgresql/json.py:376]: **mike bayer** (zzzeek) wrote:

these need to have ``..versionadded`` notes

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

File: lib/sqlalchemy/dialects/postgresql/json.py [hunk 3]
```diff
@@ -340,4 +364,28 @@ def contained_by(self, other):
...
+            return self.operate(
+                PATH_EXISTS, other, result_type=sqltypes.Boolean
+            )
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/postgresql/json.py:379]: **mike bayer** (zzzeek) wrote:

i was sort of hoping we had a more central place to do a coercion like this but it seems that we don't for a dialect-specific thing like this, will see how this goes.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

the docstrings need `..versionadded` is the main thing, otherwise is fine

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

* doc/build/changelog/unreleased_20/7147.rst (line 7): IMO for the bulleted list we dont need ; or . at the end of each line

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Tharfor the review, will update 

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

File: lib/sqlalchemy/dialects/postgresql/json.py (hunk 2 above)
Comment (sqla-tester) [reply lib/sqlalchemy/dialects/postgresql/json.py:376]: **Federico Caselli** (CaselIT) wrote:

Ack

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

updated

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4336

* doc/build/changelog/unreleased_20/7147.rst (line 7): Ack

**Pull request closed without merging**